### PR TITLE
fix (osgi) : Fix OSGi startup exceptions while using KubernetesClient/OpenShiftClient

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

- Remove ExtensionAdapter reference from ManagedKubernetesClient, no provider is exposed for ExtensionAdapter interface in kubernetes-client module. `bindExtensionAdapter` and `bindExtensionAdapter` are throwing `NullPointerException` during startup.
- Add `Provide-Capability` field in `openshift-client` and other extensions bundle configurations in order to resolve SPIFly related `NPE` (reported in #3890)

### See also
- https://aries.apache.org/documentation/modules/spi-fly.html ([#providers](https://aries.apache.org/documentation/modules/spi-fly.html#_providers))

## Type of change

 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [X] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

File: openshift-client/pom.xml [hunk 1]
```diff
@@ -33,6 +33,11 @@
       osgi.extender;
       filter:="(osgi.extender=osgi.serviceloader.registrar)"
     </osgi.require-capability>
+    <osgi.provide-capability>
+      osgi.serviceloader;
+      osgi.serviceloader=io.fabric8.kubernetes.client.extension.ExtensionAdapter,
+      io.fabric8.kubernetes.client.ServiceToURLProvider
+    </osgi.provide-capability>
```
Comment (manusa) [inline openshift-client/pom.xml:40]: This might need to be added to any extension that provides SPI capabilities.

**Review by manusa — APPROVED** [review]

**Comment by spolti** [comment]
Hi @manusa @rohanKanojia, We've found a issue with quarkus-kubernetes-client (version 2.14.0.CR1) where a Provider Not found issue is happening: 

```
ERROR: Failed to start application (with profile prod)
java.util.ServiceConfigurationError: io.fabric8.kubernetes.client.extension.ExtensionAdapter: Provider io.fabric8.knative.client.KnativeExtensionAdapter not found
```

Do you guys think that it is related with this change?
Note: Adding a BuildProducer step, apparently has fixed the issue.

```java
    @BuildStep
    void processKnative(BuildProducer<ServiceProviderBuildItem> spb) {
        spb.produce(new ServiceProviderBuildItem("io.fabric8.kubernetes.client.extension.ExtensionAdapter", KnativeExtensionAdapter.class.getName()));
    }
```

Is this the right way to have the knative client properly working in native mode, or should this fix be in the quarkus-kubernetes-client?


Thanks.

**Comment by rohanKanojia** [comment]
@spolti : I'm not entirely sure. Could you please provide complete stacktrace? As far as I remember this error was coming from SPIFly. Does it happen in your case as well?

**Comment by spolti** [comment]
Hi @rohanKanojia, here us the stacktrace:
```
java.util.ServiceConfigurationError: io.fabric8.kubernetes.client.extension.ExtensionAdapter: Provider io.fabric8.knative.client.KnativeExtensionAdapter not found
	at java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1219)
	at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:114)
	at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at io.fabric8.kubernetes.client.impl.Adapters.discoverServices(Adapters.java:106)
	at io.fabric8.kubernetes.client.impl.Adapters.<init>(Adapters.java:51)
	at io.fabric8.kubernetes.client.impl.BaseClient.<init>(BaseClient.java:102)
	at io.fabric8.kubernetes.client.impl.KubernetesClientImpl.<init>(KubernetesClientImpl.java:191)
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at io.fabric8.kubernetes.client.KubernetesClientBuilder.build(KubernetesClientBuilder.java:79)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:98)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:60)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:56)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:48)
	at org.kie.kogito.addons.quarkus.k8s.KubeDiscoveryConfigSourceInterceptor.<init>(KubeDiscoveryConfigSourceInterceptor.java:42)
	at org.kie.kogito.addons.quarkus.k8s.config.ServiceDiscoveryConfigBuilder.configBuilder(ServiceDiscoveryConfigBuilder.java:27)
	at io.quarkus.runtime.configuration.ConfigUtils.configBuilder(ConfigUtils.java:197)
	at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
	at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:110)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:69)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:122)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
```

I am not sure what SPIFly is, but in my case, it is happening with custom service discovery that uses quarkus-kubernetes-client and the io.fabric8:knative-client.

In the current Quarkus version this issue is not happening, only with the 2.14.0.CR1 that uses the fabric8 6.2.

**Comment by manusa** [comment]
This has to do with SPI and not OSGi.

Have you reported this in the Quarkus repo?
We should tackle this there.

> Note: Adding a BuildProducer step, apparently has fixed the issue.

I'm assuming that the issue is happening in native mode then. In this case we need to register any SPI provided implementations following the same pattern you showed.

**Comment by manusa** [comment]
The thing is that this is for the Knative extension, which AFAIR is not part of the Quarkus client distribution.
It's probably better to open the issue in the Quarkus Repo to discuss where the responsibility to declare the Service Provider Build Item lies, which seems like a gray area for me.

**Comment by spolti** [comment]
sure, thanks @manusa.

**Pull request merged**